yksgi/src/lib.rs

```diff
@@ -149,6 +168,271 @@ impl SGInterp {
+            nextframe = unsafe { currframe.offset(-isize::try_from(rec.size).unwrap() - 8) };
+
+            // WRITE CALLEE-SAVED REGISTERS
+            // Now we write any callee-saved registers onto the new stack. Note, that if we have
+            // pushed RBP above (which includes adjusting RBP) we need to temporarily re-adjust our
+            // pointer. This is because the CSR index calculates from the bottom of the frame, not
+            // from RBP. For example, a typical prologue looks like this:
+            //   push rbp
+            //   mov rbp, rsp
+            //   push rbx     # this has index -2
+            //   push r14     # this has index -3
+            if i > 0 {
+                for (reg, idx) in &pinfo.csrs {
+                    let mut tmp = unsafe { rbp.offset(isize::try_from(idx * 8).unwrap()) };
+                    if pinfo.hasfp {
+                        tmp = unsafe { tmp.offset(8) };
+                    }
+                    let val = registers[usize::from(*reg)];
+                    unsafe { ptr::write(tmp as *mut u64, val) };
+                }
+            }
+
+            // WRITE STACKMAP LOCATIONS.
+            // Now write all live variables to the new stack in the order they are listed in the
+            // AOT stackmap call.
+            let smcall = unsafe { get_stackmap_call(frame.pc) };
+            for (j, lv) in rec.live_vars.iter().enumerate() {
+                // Adjust the operand index by 2 to skip stackmap ID and shadow bytes.
+                let op = smcall.get_operand(u32::try_from(j+2).unwrap());
+                let val = if let Some(val) = frame.get(&op) {
+                    val.val
+                } else {
+                    todo!()
+                    //println!("probably return pointer which will disappear");
+                    //continue;
+                };
+                let l = if lv.len() == 1 {
+                    lv.get(0).unwrap()
+                } else {
+                    todo!("Deal with multi register locations");
+                };
+
+                // Iterate over all locations. Register locations just update the current value in
+                // the registers vector. Direct locations make up part of the stack so need to
+                // written to the allocated memory. Other locations we haven't encountered yet, so
+                // will deal with them as they appear.
+                match l {
+                    SMLocation::Register(reg, _size) => {
+                        registers[usize::from(*reg)] = val;
+                    }
+                    SMLocation::Direct(reg, off, _) => {
+                        if i == 0 {
+                            // skip first frame
+                            continue;
+                        }
+                        debug_assert!(op.is_alloca());
+
+                        // The sizes reported by the stackmap aren't always correct. But we can get
```

Comment:
Shouldn't we fix stackmaps then? [Not in this PR, but in general.]

Reply:
I'm not sure it's worth the effort right now to be honest.

Reply:
Maybe we should raise an issue? At the moment it looks like we're doing work at run-time that we *could* semi-easily shift to AOT.

Reply:
https://github.com/ykjit/yk/issues/579

Reply:
Thanks!